torch/distributed/fsdp/_flat_param.py

```diff
@@ -0,0 +1,474 @@
+        """
+        params_set = set(params)
+        params_set.discard(None)
+        assert len(params_set) > 0
+        param_infos: List[ParamInfo] = []
+        numels: List[int] = []
+        shapes: List[torch.Size] = []
+        prefixed_param_names: List[str] = []
+        shared_param_infos: List[SharedParamInfo] = []
+        shared_param_memo: Dict[nn.Parameter, Tuple[nn.Module, str, str]] = {}
+        params_to_flatten: List[nn.Parameter] = []
+        dtype: Optional[torch.dtype] = None
+        requires_grad: Optional[bool] = None
+        for submodule_name, submodule in module.named_modules():
+            for param_name, param in submodule.named_parameters(recurse=False):
+                if param not in params_set:
+                    continue
+                if param in shared_param_memo:
+                    prim_module, prim_module_name, prim_param_name = shared_param_memo[param]
+                    shared_param_infos.append(SharedParamInfo(
+                        param_name, submodule, submodule_name, prim_param_name,
+                        prim_module, prim_module_name,
+                    ))
+                else:
+                    if isinstance(param, FlatParameter):
+                        raise ValueError("`FlatParameter` does not support nesting")
+                    if dtype is not None and param.dtype != dtype:
+                        raise ValueError(
+                            "`FlatParameter` requires uniform dtype but got "
+                            f"{dtype} and {param.dtype}"
+                        )
+                    if requires_grad is not None and param.requires_grad != requires_grad:
+                        raise ValueError("`FlatParameter` requires uniform `requires_grad`")
+                    dtype = param.dtype
+                    requires_grad = param.requires_grad
+                    shared_param_memo[param] = (submodule, submodule_name, param_name)
+                    params_to_flatten.append(param)
+                    param_infos.append(ParamInfo(param_name, submodule, submodule_name))
+                    numels.append(param.numel())
+                    shapes.append(param.shape)
+                    prefixed_param_name = submodule_name + "." + param_name \
+                        if submodule_name else param_name
+                    prefixed_param_names.append(prefixed_param_name)
+        assert requires_grad is not None
+        self.flat_param = FlatParamHandle.flatten_params(params_to_flatten, requires_grad)
+        self.flat_param.init_metadata(
+            param_infos, numels, shapes, prefixed_param_names, shared_param_infos,
+        )
+
+    @staticmethod
+    def flatten_params(
+        params: Sequence[torch.Tensor],
+        requires_grad: bool,
+    ) -> FlatParameter:
+        """
+        Flattens the parameters in ``params`` into a single
+        :class:`FlatParameter`. This should be the only way used to construct
```

Comment:
We might need to do some tricks since this code is in Python:
https://stackoverflow.com/questions/8212053/private-constructor-in-python

Reply:
When I say that this should be the only way to construct `FlatParameter`s, it is more directed toward developers.

If a user wants to create his/her own `FlatParameter`, it will simply be the same as creating a normal `nn.Parameter` since `FlatParameter` does not provide any additional functionality.

I think there is not much harm if we do not **strictly** enforce everyone **must** use this method to construct `FlatParameter`s.